MatrixSDKTests/MXSessionTests.m

```diff
@@ -65,6 +67,110 @@ - (void)tearDown
+        
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+            
+            [NSThread sleepForTimeInterval:1.0];
+            
+            XCTAssertEqual(cache.syncResponseIds.count,
+                           0,
+                           @"Initial sync cache must be reset after successful initialization");
+
+            [expectation fulfill];
+            [store deleteAllData];
+        });
+    }];
+}
+
+// Check MXSession updates initial sync cache when an error occurs handling the sync response.
+//
+// - Have Bob start a new session
+// - Run initial sync on Bob's session
+// - In the middle of the process, close the session to simulate a crash
+// -> The initial sync cache must be updated
+// - Restart the session
+// -> The initial sync cache must be used
+- (void)testInitialSyncFailure
+{
+    id<MXStore> store = [[MXFileStore alloc] init];
+    
+    [matrixSDKTestsData doMXRestClientTestWithBob:self readyToTest:^(MXRestClient *bobRestClient, XCTestExpectation *expectation) {
+
+        MXCredentials *credentials = [MXCredentials initialSyncCacheCredentialsFrom:matrixSDKTestsData.bobCredentials];
+        id<MXSyncResponseStore> cache = [[MXSyncResponseFileStore alloc] initWithCredentials:credentials];
+        mxSession = [[MXSession alloc] initWithMatrixRestClient:bobRestClient];
+        
+        //  listen for session state change notification
+        __block id observer = [[NSNotificationCenter defaultCenter] addObserverForName:kMXSessionStateDidChangeNotification object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {
+            
+            if (mxSession.state == MXSessionStateRunning)
+            {
+                //  stop observing
+                [[NSNotificationCenter defaultCenter] removeObserver:observer];
+                
+                //  2. simulate an unexpected session close (like a crash)
+                [mxSession close];
+                mxSession = nil;
+                
+                XCTAssertGreaterThan(cache.syncResponseIds.count,
+                                     0,
+                                     @"Session must cache initial sync responses in case of a failure");
+                
+                //  3. recreate the session
+                mxSession = [[MXSession alloc] initWithMatrixRestClient:bobRestClient];
+
+                [mxSession setStore:store success:^{
+                    [mxSession start:^{
+                        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+
+                            [NSThread sleepForTimeInterval:1.0];
+
+                            XCTAssertEqual(cache.syncResponseIds.count,
+                                           0,
+                                           @"Initial sync cache must be used after successful restart");
+
+                            [expectation fulfill];
+                            [store deleteAllData];
```

Comment:
```suggestion
```

There is no need. Plus, it should happen before the line above.

Reply:
You right, removed.